### PR TITLE
[SecurionPay] Implemented `authorise/3`

### PR DESCRIPTION
**Pull request by xC-Stony**
Added docs for the gateway and implemented authorise function.
Authorise function can be called with either cardID and customerId or
with card details. In latter case we have to create a token and then use the token id
for authorizing the payment.

### CODE REVIEW

File: lib/gringotts/gateways/securion_pay.ex [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  To transfer the funds to merchant's account follow this up with a `capture/3`.
+
+  SecurionPay returns a `chargeId` which uniquely identifies a transaction (available in the `Response.id` field) 
+  which should be stored at your side to use in:
+
+  * `capture/3` an authorized transaction.
+  * `void/2` a transaction.
+
+  ## Example 1
+      iex> amount = Money.new(20, :USD}
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: 1. Typo `}`, and in line 81 there's a stray `:`
2. You can rewrite this section as:
```
## Examples
### With a `card_token` and `customer_token`
...
### With a `CreditCard` struct
...
```

File: lib/gringotts/gateways/securion_pay.ex [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    header = [{"Authorization", "Basic " <> opts[:config][:secret_key]}]
+    token_id = create_token(card, header)
+
+    token_id
+    |> create_params(amount, false)
+    |> commit(:post, "charges", header)
+    |> respond
+  end
+
+  def authorize(amount, card, opts) do
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: Add a guard clause like so:
```elixir
def authorize(amount, card_token, opts) when is_binary(card_token) do
  ...
end
```

File: lib/gringotts/gateways/securion_pay.ex [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  end
+
+  ###############################################################################
+  #                                PRIVATE METHODS                              #
+  ###############################################################################
+
+  # Creates the parameters for authorise function when 
+  # cardId and customerId is provided.
+  @spec create_params({}, Money.t(), boolean) :: {[]}
+  defp create_params({cardId, customerId}, amount, captured) do
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: 1. Why do you want to pass these as a tuple _to this function_?
2. Please follow the [lexmag style guide](https://github.com/lexmag/elixir-style-guide) for variable naming (snake_case only).

File: lib/gringotts/gateways/securion_pay.ex [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    {currency, value, _, _} = Money.to_integer_exp(amount)
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: Please document the fact that SecurionPay works in sub-divided unit, aka. "cents".

File: lib/gringotts/gateways/securion_pay.ex [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  places a hold on the transaction `amount` in the customer’s issuing bank and
+  also triggers risk management. Funds are not transferred.
+
+  The second argument can be a CreditCard or a cardId. The customerId of the cutomer who owns the card must be
+  given in optional field. 
+
+  To transfer the funds to merchant's account follow this up with a `capture/3`.
+
+  SecurionPay returns a `chargeId` which uniquely identifies a transaction (available in the `Response.id` field) 
+  which should be stored at your side to use in:
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: ... stored by the caller for use ...

File: lib/gringotts/gateways/securion_pay.ex [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      iex> result = Gringotts.Gateways.SecurionPay.authorize(amount, card, opts)
+
+  ## Example 2
+      iex> amount = Money.new(20, :USD}
+      iex> opts = [config: [:secret_key: "c2tfdGVzdF9GZjJKcHE1OXNTV1Q3cW1JOWF0aWk1elI6"], customerId: "cust_zpYEBK396q3rvIBZYc3PIDwT"]
+      iex> card = "card_LqTT5tC10BQzDbwWJhFWXDoP"
+      iex> result = Gringotts.Gateways.SecurionPay.authorize(amount, card, opts)
+
+  """
+  @spec authorize(Money.t(), CreditCard.t() | {}, keyword) :: {:ok | :error, Response}
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: IMO, the `authorize` spec should be, following the work in `Stripe` and `WireCard`,
```elixir
@spec authorize(Money.t(), CreditCard.t() | String.t(), keyword) :: {:ok | :error, Response.t()}
```
That is, the `card` argument can be either a `CreditCard` struct or a string (`card_token`). Other tokens, if needed, can be passed via `opts`.

File: lib/gringotts/gateways/securion_pay.ex [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  defp respond({:ok, %{status_code: 200, body: body}}) do
+    parsed_body = Poison.decode!(body)
+
+    {:ok,
+     %{
+       success: true,
+       id: Map.get(parsed_body, "id"),
+       token: Map.get(parsed_body["card"], "id"),
+       status_code: 200,
+       gateway_code: 200,
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: 1. `gateway_code` cannot be set like this. It is the gateway's response code "as-is". If there's no response code, let it be `nil`.
I'm sure SecurionPay returns an error code that can go in this field.
2. Also, you don't have to explicitly set the `nil` values in a struct.

File: lib/gringotts/gateways/securion_pay.ex [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    }
+  end
+
+  defp create_token(card, header) do
+    [
+      {"number", card.number},
+      {"expYear", card.year},
+      {"cvc", card.verification_code},
+      {"expMonth", card.month},
+      {"cardholderName", card.first_name <> " " <> card.last_name}
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: `{"cardholderName", CreditCard.full_name(card)}`

File: lib/gringotts/gateways/securion_pay.ex [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+      {"amount", value},
+      {"currency", to_string(currency)},
+      {"card", cardId},
+      {"captured", "#{captured}"},
+      {"customerId", customerId}
+    ]
+  end
+
+  # Creates the parameters for authorise when token is provided.
+  @spec create_params(Integer.t(), Money.t(), boolean) :: {[]}
```
Comment (oyeb) [inline lib/gringotts/gateways/securion_pay.ex]: 1. Bad `spec`.
2. Both `create_params` clauses have duplicated code (amount fields), that can be pulled out.

File: lib/gringotts/gateways/securion_pay.ex (hunk 4 above)
Comment (xC-Stony) [reply lib/gringotts/gateways/securion_pay.ex]: Can I document the fact in the module docs that all transactions take place in minor units?

File: lib/gringotts/gateways/securion_pay.ex [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+  which should be stored by the caller for using in:
+
+  * `capture/3` an authorized transaction.
+  * `void/2` a transaction.
+
+  ## Example
+  ### With a `CreditCard` struct
+      iex> amount = Money.new(20, :USD)
+      iex> opts = [config: [secret_key: "c2tfdGVzdF9GZjJKcHE1OXNTV1Q3cW1JOWF0aWk1elI6"]]
+      iex> card = %CreditCard{
```
Comment (aashishsingh2803) [inline lib/gringotts/gateways/securion_pay.ex]: Follow proper code format for example.

File: lib/gringotts/gateways/securion_pay.ex [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+  # Parses SecurionPay's response and returns a `Gringotts.Response` struct
+  # in a `:ok`, `:error` tuple.
+  @spec respond(term) :: {:ok | :error, Response}
+
+  defp respond({:ok, %{status_code: 200, body: body}}) do
+    parsed_body = Poison.decode!(body)
+
+    {:ok,
+     %{
+       success: true,
```
Comment (aashishsingh2803) [inline lib/gringotts/gateways/securion_pay.ex]: Respond function should return response of predefined format as in Gringotts.Response.ex

**Pull request closed without merging**